src/util/ConfigManager/ConfigOption.h

```diff
@@ -247,7 +257,72 @@ class ConfigOption {
     }
   }
 
+  /*
+  @brief Add a function to the configuration option for verifying, that the
+  value of the configuration option is valid. Will always be called after
+  setting the value of the configuration option.
+
+  @tparam ValidatorFunction A function, that takes a single argument of type
+  `const T&`, with `T` being the type of value, this config options holds, and
+  returns a bool.
+
+  @param validatorFunction Checks, if the value of the configuration option is
+  valid. Should return true, if it is valid.
+  @param errorMessage A `std::runtime_error` with this as an error message will
+  get thrown, if the `validatorFunction` returns false.
+ */
+  template <isValidatorWithSingleParameterTypeOutOfVariant<AvailableTypes>
+                ValidatorFunction>
+  void addValidator(ValidatorFunction validatorFunction,
+                    const std::string& errorMessage) {
+    /*
+    Deducing the exact parameter of a given callable object is a major hassle
+    and neigh impossible. Instead we just use the internal type.
+    */
+    std::visit(
+        [this, &validatorFunction,
+         &errorMessage]<typename T>(const Data<T>& data) {
+          /*
+          Can `validatorFunction` be called with the type of this
+          configuration option?
+          Note: This also allows implicit conversion. For example, a function
+          that takes an int, would work for a configuration option that holds a
+          bool.
+          */
+          if constexpr (Validator<ValidatorFunction, const T&>) {
+            /*
+            Add a function wrapper to our list of validators, that calls the
+            `validatorFunction` with the value and throws an exception, if the
+            `validatorFunction` returns false.
+            */
+            validators_.emplace_back([validatorFunction,
+                                      valuePointer = data.variablePointer_,
+                                      errorMessage]() mutable {
+              AD_CORRECTNESS_CHECK(valuePointer != nullptr);
+
+              if (!std::invoke(validatorFunction, *valuePointer)) {
+                throw std::runtime_error(errorMessage);
+              } else {
+                return true;
+              }
+            });
+          } else {
+            throw std::runtime_error(absl::StrCat(
+                "Adding of validator to configuration option '", identifier_,
+                "' failed. Configuration option holds value of type '",
+                getActualValueTypeAsString(),
+                "', yet validator function is not invocable with a parameter "
+                "of this type."));
```

Comment:
I think you should make this a member function of the proxy, then you can use the same typesafe checks as for the other functionality (maybe you can even factor out some of the common code, as this is basically `ConfigManager::addOption` but with only one option and the result is stored in the parameter.

Maybe you should therefore make the proxy a `friend` of the `ConfigOption` to let it access the type-erased `validators`.

Reply:
I don't think, this is a good idea. It would hide functionality of the class *inside* a different class. Or more specifically, it would hide the access.
What's the point of having a `ConfigOption` class, if we outsource its member functions?

Reply:
We should discuss this more in person.